lib/spack/spack/bootstrap/core.py

```diff
@@ -493,6 +513,49 @@ def _add_externals_if_missing() -> None:
     )
 
 
+class BootstrapResource():
+    """Represents a resource required by Spack to run, fetched as part
+    of the bootstrapping operation
+
+    More or less identical to package based resources but exposes additional logic related
+    to bootstrapping/ resource visibility.
+
+    Composes a name and a fetch strategy
+    """
+    def __init__(self, name, conf):
+        """
+        Args:
+            name (str): Name of resource to be fetched
+            conf (dict): Dictionary representing resource endpoint layout
+        """
+        self._name = name
+        fetcher = spack.fetch_strategy.URLFetchStrategy(url=conf["endpoint"], checksum=conf["sha256"])
+        resource = spack.resource.Resource(name, fetcher, destination=self.stage.path)
+        stage = spack.stage.Stage(fetcher, path=str(_windows_resource_root()))
+        self.stage = spack.stage.ResourceStage(fetcher, stage, resource, path=str(_windows_resource_root()))
+
+    def acquire_resource(self):
+        "fetches, expands, and 'installs' resource"
+        with self.stage as s:
+            s.fetch()
+            s.expand_archive()
+            shutil.copytree(pathlib.Path(s.path) / self._name, _windows_resource_root() / self._name)
+
+
+def ensure_resource(name):
+    """Acquires resource from configured sources"""
+    resource_sources = spack.config.get("resource:sources")
+    resources = []
+    for source in resource_sources:
+        with open(pathlib.Path(source["resource_layout"]) / "layout.yaml", "r") as f:
+            cnf = spack.util.spack_yaml.load(f)
+            resources.append(BootstrapResource(name, cnf))
+    for resource in resources:
+        if resource.acquire_resource():
+            insert_resource_into_environment(name)
+            return
```

Comment:
It seems like there's a disconnect: you call

```
    ensure_resource("file")
    ensure_resource("gpg")
```

but in this context, the resource is the `layout.yaml` that stores multiple sources, which all appear to be unpacked here.

This is partially exacerbated by reuse of the term `resource` (i.e. both `BootstrapResource`, the existing `Resource` object in Spack, and calling the the endpoints `resources` in this function (can you create more-distinct names?).

Reply:
You're right there is. This method needs a refactor, the interface of calling `ensure_resource` is what I had intended, but the resource gathering doesn't consider that. 
I'll refactor (and rename)